### PR TITLE
Implement basic login

### PR DESCRIPTION
**Pull request by thanhconguet214**
ready

### CODE REVIEW

**Comment by thanhconguet214** [comment]
ready

**Comment by thanhconguet214** [comment]
ready

File: Gemfile [hunk 1]
```diff
@@ -4,6 +4,8 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 ruby "2.5.1"
 gem "rails", "~> 5.2.0"
 gem "bootstrap-sass", "~> 3.3.7"
+gem "bootstrap-will_paginate", "1.0.0"
+gem "sprockets-rails", :require => "sprockets/railtie"
```
Comment (huuphuc93) [inline Gemfile]: 2 gem này dùng làm gì em? sao phải thêm vào?

File: app/assets/javascripts/sessions.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (huuphuc93) [inline app/assets/javascripts/sessions.coffee]: remove this file

File: app/assets/stylesheets/sessions.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
```
Comment (huuphuc93) [inline app/assets/stylesheets/sessions.scss]: remove this file

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+  	user = User.find_by email: params[:session][:email].downcase
```
Comment (huuphuc93) [inline app/controllers/sessions_controller.rb]: indent

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class SessionsController < ApplicationController
+  def new
```
Comment (huuphuc93) [inline app/controllers/sessions_controller.rb]: def new; end

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -2,10 +2,9 @@ class UsersController < ApplicationController
 
   def show
     @user = User.find_by id: params[:id]
-    return if @user
+      return if @user
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: indent

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+module SessionsHelper
+  def log_in(user)
```
Comment (huuphuc93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/models/user.rb [hunk 8]
```diff
@@ -7,4 +7,10 @@ class User < ApplicationRecord
     uniqueness: {case_sensitive: false}
   validates :password, presence: true, length: {minimum: Settings.minimum}
   has_secure_password
+
+  def User.digest(string)
```
Comment (huuphuc93) [inline app/models/user.rb]: use class < self for class method

File: app/views/layouts/_header.html.erb [hunk 9]
```diff
@@ -1,13 +1,29 @@
 <header class="navbar navbar-fixed-top navbar-inverse">
   <div class="container">
     <%= link_to "sample app", root_path, id: "logo" %>
-    <nav>
-      
+    <nav>      
       <ul class="nav navbar-nav navbar-right">
         <li><%= link_to "Home",   root_path %></li>
```
Comment (huuphuc93) [inline app/views/layouts/_header.html.erb]: use 1 space between "Home",  and  root_path

**Comment by huuphuc93** [comment]
commented

**Comment by thanhconguet214** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 10]
```diff
@@ -0,0 +1,19 @@
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user&.authenticate(params[:session][:password])
```
Comment (mdmahmudurrahman) [inline app/controllers/sessions_controller.rb]: Em, remove optional ()

File: app/models/user.rb [hunk 11]
```diff
@@ -7,4 +7,12 @@ class User < ApplicationRecord
     uniqueness: {case_sensitive: false}
   validates :password, presence: true, length: {minimum: Settings.minimum}
   has_secure_password
+
+  class << self
+    def digest(string)
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Remove optional ()

File: app/views/layouts/_header.html.erb [hunk 12]
```diff
@@ -1,13 +1,29 @@
...
-      
+    <nav>      
       <ul class="nav navbar-nav navbar-right">
-        <li><%= link_to "Home",   root_path %></li>
-        <li><%= link_to "Help",   help_path %></li>
-        <li><%= link_to "Log in", '#' %></li>
+        <li><%= link_to "Home", root_path %></li>
+        <li><%= link_to "Help", help_path %></li>
+        <% if logged_in? %>
+          <li><%= link_to "Users", '#' %></li>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_header.html.erb]: Please use double quote instead of using single quote.

File: app/views/layouts/_header.html.erb [hunk 13]
```diff
@@ -1,13 +1,29 @@
...
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              Account <b class="caret"></b>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_header.html.erb]: I18n

File: app/views/layouts/_header.html.erb [hunk 14]
```diff
@@ -1,13 +1,29 @@
...
+            </a>
+            <ul class="dropdown-menu">
+              <li><%= link_to "Profile", current_user %></li>
+              <li><%= link_to "Settings", '#' %></li>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_header.html.erb]: Please use double quote

File: app/views/layouts/_header.html.erb [hunk 15]
```diff
@@ -1,13 +1,29 @@
...
+              <li class="divider"></li>
+              <li>
+                <%= link_to "Log out", logout_path, method: :delete %>
+              </li>
+            </ul>
+          </li>
+        <% else %>
+        <li><%= link_to "Log in", login_path %></li>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_header.html.erb]: 1. Add 2 space indentation as this line is nested inside a block
2. I18n

File: app/views/sessions/new.html.erb [hunk 16]
```diff
@@ -0,0 +1,14 @@
+<% provide :title, t(".login") %>
+<h1>Log in</h1>
```
Comment (mdmahmudurrahman) [inline app/views/sessions/new.html.erb]: I18n

File: app/views/sessions/new.html.erb [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:session, url: login_path) do |f| %>
```
Comment (mdmahmudurrahman) [inline app/views/sessions/new.html.erb]: Please remove optional ()

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by thanhconguet214** [comment]
ready

File: app/models/user.rb [hunk 18]
```diff
@@ -7,4 +7,12 @@ class User < ApplicationRecord
     uniqueness: {case_sensitive: false}
   validates :password, presence: true, length: {minimum: Settings.minimum}
   has_secure_password
+
+  class << self
+    def digest string
+      cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+        BCrypt::Engine.cost
+        BCrypt::Password.create(string, cost: cost)
```
Comment (huuphuc93) [inline app/models/user.rb]: indent

**Comment by thanhconguet214** [comment]
ready

File: app/views/sessions/new.html.erb [hunk 19]
```diff
@@ -0,0 +1,14 @@
+<% provide :title, t(".login") %>
+<h1><%= t(".login") %></h1>
```
Comment (mdmahmudurrahman) [inline app/views/sessions/new.html.erb]: Please remove optional ()

File: app/views/sessions/new.html.erb [hunk 20]
```diff
@@ -0,0 +1,14 @@
...
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for :session, url: login_path do |f| %>
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+      <%= f.submit t(".login"), class: "btn btn-primary" %>
+    <% end %>
+    <p> <%= t ".newuser" %><%= link_to t(".sign"), signup_path %></p>
```
Comment (mdmahmudurrahman) [inline app/views/sessions/new.html.erb]: Remove unnecessary white space.
`<p><%= t ".newuser" %><%= link_to t(".sign"), signup_path %></p>`

File: config/routes.rb [hunk 21]
```diff
@@ -1,9 +1,13 @@
 Rails.application.routes.draw do
+  get "sessions/new"
```
Comment (mdmahmudurrahman) [inline config/routes.rb]: Do you still need this redundant routing? If so, then why ?

File: app/models/user.rb [hunk 22]
```diff
@@ -7,4 +7,12 @@ class User < ApplicationRecord
     uniqueness: {case_sensitive: false}
   validates :password, presence: true, length: {minimum: Settings.minimum}
   has_secure_password
+
+  class << self
+    def digest string
+      cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+      BCrypt::Engine.cost
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Add 2 space indentation here as you break previous line

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by thanhconguet214** [comment]
ready

**Comment by mdmahmudurrahman** [comment]
commented
This (pr) can be (merged)
CC: [To:2076564] Nguyen Huy Hung C (93)
CC: [To:2612902] Mai Huu Phuc

**Pull request merged**